cartographer_ros/cartographer_ros/node.cc

```diff
@@ -534,21 +513,58 @@ cartographer_ros_msgs::StatusResponse Node::FinishTrajectoryUnderLock(
 bool Node::HandleStartTrajectory(
     ::cartographer_ros_msgs::StartTrajectory::Request& request,
     ::cartographer_ros_msgs::StartTrajectory::Response& response) {
-  absl::MutexLock lock(&mutex_);
-  TrajectoryOptions options;
-  if (!FromRosMessage(request.options, &options) ||
-      !ValidateTrajectoryOptions(options)) {
+  const auto trajectory_states = map_builder_bridge_.GetTrajectoryStates();
+  if (!(trajectory_states.count(request.to_trajectory_id)) ||
+      trajectory_states.at(request.to_trajectory_id) ==
+          TrajectoryState::DELETED) {
+    const std::string error =
+        absl::StrCat("Trajectory ", std::to_string(request.to_trajectory_id),
+                     " doesn't exist or has been deleted.");
+    LOG(ERROR) << error;
+    response.status.code = cartographer_ros_msgs::StatusCode::INVALID_ARGUMENT;
+    response.status.message = error;
+    return true;
+  }
+
+  TrajectoryOptions trajectory_options;
+  std::tie(std::ignore, trajectory_options) = LoadOptions(
+      request.configuration_directory, request.configuration_basename);
+
+  const auto pose = ToRigid3d(request.initial_pose);
+  if (request.use_initial_pose) {
+    if (!pose.IsValid()) {
+      const std::string error =
+          "Invalid pose argument. Orientation quaternion must be normalized.";
+      LOG(ERROR) << error;
+      response.status.code =
+          cartographer_ros_msgs::StatusCode::INVALID_ARGUMENT;
+      response.status.message = error;
+      return true;
+    }
+
+    ::cartographer::mapping::proto::InitialTrajectoryPose
+        initial_trajectory_pose;
+    initial_trajectory_pose.set_to_trajectory_id(request.to_trajectory_id);
+    *initial_trajectory_pose.mutable_relative_pose() =
+        cartographer::transform::ToProto(pose);
+    initial_trajectory_pose.set_timestamp(cartographer::common::ToUniversal(
+        ::cartographer_ros::FromRos(ros::Time(0))));
+    *trajectory_options.trajectory_builder_options
+         .mutable_initial_trajectory_pose() = initial_trajectory_pose;
+  }
+
+  if (!ValidateTrajectoryOptions(trajectory_options)) {
     const std::string error = "Invalid trajectory options.";
     LOG(ERROR) << error;
     response.status.code = cartographer_ros_msgs::StatusCode::INVALID_ARGUMENT;
     response.status.message = error;
-  } else if (!ValidateTopicNames(request.topics, options)) {
-    const std::string error = "Invalid topics.";
+  } else if (!ValidateTopicNames(trajectory_options)) {
+    const std::string error = "Topics are already used by another trajectory.";
```

Comment:
same here. Store the error message in response.status.message and log this instead.